pkg/controller/orchestrator/orchestrator_reconcile.go

```diff
@@ -0,0 +1,478 @@
+// unregistered nodes and to remove nodes that does not exists.
+func (ou *orcUpdater) registerUnregisterNodesInOrc(instances InstancesSet) InstancesSet {
+	var (
+		// hosts that should be discovered
+		shouldBeDiscovered []string
+		// list of instances from orchestrator that should be removed
+		shouldBeRemoved []string
+		// the list of instances that are in orchestrator and k8s
+		instancesFiltered InstancesSet
+	)
+
+	for i := 0; i < ou.cluster.Status.ReadyNodes; i++ {
+		host := ou.cluster.GetPodHostname(i)
+		if inst := instances.GetInstance(host); inst == nil {
+			// host is not present into orchestrator
+			// register new host into orchestrator
+			shouldBeDiscovered = append(shouldBeDiscovered, host)
+		} else {
+			// this instance is present in both k8s and orchestrator
+			instancesFiltered = append(instancesFiltered, *inst)
+		}
+	}
+
+	// remove all instances from orchestrator that does not exists in k8s
+	for _, inst := range instances {
+		if i := instancesFiltered.GetInstance(inst.Key.Hostname); i == nil {
+			shouldBeRemoved = append(shouldBeRemoved, inst.Key.Hostname)
+		}
+	}
+
+	ou.discoverNodesInOrc(shouldBeDiscovered)
+	ou.forgetNodesFromOrc(shouldBeRemoved)
+
+	return instancesFiltered
+}
+
+func (ou *orcUpdater) discoverNodesInOrc(hosts []string) {
+	for _, host := range hosts {
+		if err := ou.orcClient.Discover(host, mysqlPort); err != nil {
+			log.Error(err, "failed to register host with orchestrator", "host", host)
+		}
+	}
+}
+
+func (ou *orcUpdater) forgetNodesFromOrc(hosts []string) {
+	for _, host := range hosts {
+		if err := ou.orcClient.Forget(host, mysqlPort); err != nil {
+			log.Error(err, "failed to forget host with orchestrator", "host", host)
+		}
+	}
+}
+
+func (ou *orcUpdater) getRecoveriesToAck(recoveries []orc.TopologyRecovery) []orc.TopologyRecovery {
+	toAck := []orc.TopologyRecovery{}
+
+	if len(recoveries) == 0 {
+		return toAck
+	}
+
+	i, found := condIndexCluster(ou.cluster, api.ClusterConditionReady)
+	if !found || ou.cluster.Status.Conditions[i].Status != core.ConditionTrue {
+		log.Info("cluster not ready for acknowledge", "cluster", ou.cluster)
+		return toAck
+	}
+
+	timeSinceReady := time.Since(ou.cluster.Status.Conditions[i].LastTransitionTime.Time).Minutes()
+	if timeSinceReady < healtyMoreThanMinutes {
+		log.Info("cluster not ready for acknowledge", "timeSinceReady", timeSinceReady, "trashold", healtyMoreThanMinutes)
+		return toAck
+	}
+
+	for _, recovery := range recoveries {
+		if !recovery.Acknowledged {
+			// skip if it's a new recovery, recovery should be older then <healtyMoreThanMinutes> minutes
+			startTime, err := time.Parse(time.RFC3339, recovery.RecoveryStartTimestamp)
+			if err != nil {
+				log.Error(err, "time parse error",
+					"recovery", recovery,
+				)
+				continue
+			}
+			if time.Since(startTime).Minutes() < healtyMoreThanMinutes {
+				// skip this recovery
+				log.Error(nil, "tries to recover to sson", "recovery", recovery)
```

Comment:
This is not an error, so we should `log.V(2).Info`. Also the message should read `skip acknowledging recovery since grace period has not passed`. Also log `recoveryGraceMinutes`.